### PR TITLE
[seekhandler] get rid of inaccurate seek size percent calculation

### PR DESCRIPTION
**Pull request by xhaggi**
This removes the inaccurate seek size calculation which relies on percent and uses seconds instead.
As we need the seek size percent for the GUI seek progress bar this calculation is refactored into our CGUIInfoManager.

@FernetMenta mind taking a look

### CODE REVIEW

File: xbmc/utils/SeekHandler.cpp [hunk 1]
```diff
@@ -130,56 +128,44 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
...
+    if( duration )
+      speed *= duration;
+    else
+      speed /= g_infoManager.GetFPS();
 
-      if (forward)
-        m_percent += amount * amount * speed;
-      else
-        m_percent -= amount * amount * speed;
+    int seekSize = (amount * amount * speed) * g_infoManager.GetTotalPlayTime() / 100;
```
Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: seekhandler should not be dependent on infomanager. the values in infomanager are for display only and should not serve for any kind of behaviour.

Comment (xhaggi) [inline xbmc/utils/SeekHandler.cpp]: agree, we should use `CApplication::GetTotalTime()` instead.

**Comment by xhaggi** [comment]
> seekhandler should not be dependent on infomanager.

@FernetMenta most of the work is done in the fixup commit but there are 3 calls to the CGUIInfoManager i can't remove ATM.

File: xbmc/utils/SeekHandler.cpp [hunk 2]
```diff
@@ -130,56 +128,46 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
...
-      if( duration )
-        speed *= duration;
-      else
-        speed /= g_infoManager.GetFPS();
+    //100% over 1 second.
+    float speed = 100.0f;
+    if( duration )
+      speed *= duration;
+    else
+      speed /= g_infoManager.GetFPS();
```
Comment (xhaggi) [inline xbmc/utils/SeekHandler.cpp]: this one

**Comment by xhaggi** [comment]
this one
https://github.com/xbmc/xbmc/pull/7012/files#diff-c647c68d33c198547ac05b3c7e0eafc4R293
and 
https://github.com/xbmc/xbmc/pull/7012/files#diff-c647c68d33c198547ac05b3c7e0eafc4R299

**Comment by xhaggi** [comment]
i'm not sure if the calls to `g_infoManager.SetDisplayAfterSeek()` are really necessary.

File: xbmc/utils/SeekHandler.cpp (hunk 2 above)
Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: graphicsContext is master of FPS

File: xbmc/utils/SeekHandler.cpp [hunk 3]
```diff
@@ -130,56 +128,46 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
...
 
-      if (forward)
-        m_percent += amount * amount * speed;
-      else
-        m_percent -= amount * amount * speed;
+    double totalTime = g_application.GetTotalTime();
```
Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: this is not correct either. g_application is not master of total time, player is.
but why is this needed? either instruct player to seek to a given absolute time or relative. in neither case seek hander needs to know total time.

File: xbmc/utils/SeekHandler.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/utils/SeekHandler.cpp]: sure it makes sense to handle it in player, but imo we should do one at a time.
ATM i want to get rid of the percent calculation and most of the CGUIInfoManager dependencies.

Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: how often do you want to touch it? it is still broken for material having no total playing time, hence all those changes are fixes not improvements. I see no reason why we should continue to do it wrong. let's eliminate the flaws which are the dependency to infomanager and lack of relative seek capability of player.

Comment (xhaggi) [inline xbmc/utils/SeekHandler.cpp]: if you look at the fixup commit you see that `GetTotalTime()` is only used for analog seeking same as before we introduced the additive seeking. We need to change the analog seek size calculation to get rid of it. Do you have an idea how?

File: xbmc/utils/SeekHandler.cpp (hunk 2 above)
Comment (xhaggi) [inline xbmc/utils/SeekHandler.cpp]: thanks will change it.

File: xbmc/utils/SeekHandler.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: > if you look at the fixup commit you see that GetTotalTime() s only used for analog seeking

ahh, I missed this. thanks, then it's good for now

**Comment by xhaggi** [comment]
I've added a new commit which adds a new parameter to `CApplication::SeekTime()` to handle relative seeking. ATM I can't adjust the Player itself because of the stack stuff in `CApplication::SeekTime()` I'm not very familiar with.

@FernetMenta a second look would be good.

**Comment by FernetMenta** [comment]
I would have created separate methods for relative seeking, this way you can leave the stacking stuff as is and don't need many if blocks in a single Seek method. Assume you have stacked 2 files and we are 10 sec close to the end of the first. any relative seek forward > 10 sec should stop at the end of file 1.

**Comment by xhaggi** [comment]
@FernetMenta did you looked at the last commit? there is only a single condition added to `CApplication::SeekTime()` to do the relative calculation. Why do you think we need a separate method except for convenience reason?

**Comment by FernetMenta** [comment]
> Why do you think we need a separate method except for convenience reason?

see my last comment, i.e. seeking to an absolute time has to consider stacking, relative does not. there may be more reasons we currently don't think of. imo it is much cleaner and clearer having separate methods.

**Comment by FernetMenta** [comment]
another reason is accurate seeks which should be switched off by default. player will get a separate method too.

**Comment by xhaggi** [comment]
@FernetMenta okay but I'm not sure how this should be done. It would be great if you could help me a bit with the player related stuff. It speed up things because I never touched that part.

**Comment by FernetMenta** [comment]
can you have a look at this: https://github.com/FernetMenta/xbmc/commit/8b33a426595446fa87cd03f6e73a71b9968cc3ec

**Comment by xhaggi** [comment]
great thx will cherry pick it tomorrow.

**Comment by xhaggi** [comment]
@FernetMenta updated on top of your player changes. mind taking a look again.

File: xbmc/Application.h [hunk 4]
```diff
@@ -297,6 +297,11 @@ class CApplication : public CXBApplicationEx, public IPlayerCallback, public IMs
 #endif
 
   CApplicationPlayer* m_pPlayer;
+  /*!
+   \brief Returns the application player
+   \return The application player
+   */
+  CApplicationPlayer* GetPlayer() { return m_pPlayer; }
```
Comment (FernetMenta) [inline xbmc/Application.h]: m_pPlayer is not supposed to be exposed. why do you access the internals? applicationplayer has a method SeekTimeRelative

File: xbmc/utils/SeekHandler.cpp [hunk 5]
```diff
@@ -228,17 +202,14 @@ void CSeekHandler::Process()
...
-
-    // reset seek step size
-    g_infoManager.SetSeekStepSize(0);
+    CSingleLock lock(m_critSection);
+    m_performingSeek = true;
 
-    // calculate the seek time
-    double time = g_infoManager.GetTotalPlayTime() * m_percent * 0.01;
+    // perform relative seek
+    g_application.GetPlayer()->SeekTimeRelative(static_cast<int64_t>(m_seekSize * 1000));
```
Comment (FernetMenta) [inline xbmc/utils/SeekHandler.cpp]: g_application.m_pPlayer->

File: xbmc/Application.h (hunk 4 above)
Comment (xhaggi) [inline xbmc/Application.h]: will drop it

**Comment by xhaggi** [comment]
@FernetMenta done

**Comment by FernetMenta** [comment]
I will test it this evening if I get time.
btw: I flagged this as a fix because additive seeking enabled accurate seeks which are expensive. this pr fixes this and makes step forward/backward fast again.

File: xbmc/GUIInfoManager.cpp [hunk 6]
```diff
@@ -4126,6 +4124,23 @@ int CGUIInfoManager::GetPlayTimeRemaining() const
   return iReverse > 0 ? iReverse : 0;
 }
 
+float CGUIInfoManager::GetSeekPercent() const
+{
+  if (g_infoManager.GetTotalPlayTime() == 0)
+    return 0.0f;
+
+  float percentPlayTime = static_cast<float>(g_infoManager.GetPlayTime()) / g_infoManager.GetTotalPlayTime() * 0.1f;
```
Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: why the g_infoManager prefix? c/p ?

**Comment by FernetMenta** [comment]
works good here.

File: xbmc/GUIInfoManager.cpp (hunk 6 above)
Comment (xhaggi) [inline xbmc/GUIInfoManager.cpp]: yep ;) .. will fix it asap

**Comment by xhaggi** [comment]
jenkins build and merge

File: xbmc/GUIInfoManager.cpp (hunk 6 above)
Comment (xhaggi) [inline xbmc/GUIInfoManager.cpp]: done
